### PR TITLE
Wide slice

### PR DESCRIPTION
**Pull request by anicolao**
A first pass at introducing `ExperimentalCollapseOptions` so that `experimentalAddMove` can have the behaviours a roux solver would need.

I stopped short of pushing these options up to the element parameters so that we can check the shape of it before I change the public facing API. It isn't clear to me whether we'd prefer to keep supporting `coalesce` at the element level (it would mean sameDirection=oppositeDirection=true) and just add support for `sameDirection` and `oppositeDirection` or if we'd prefer to remove coalesce from the API and update all call sites.

### CODE REVIEW

**Review by lgarron — COMMENTED** [review]
Thanks for the PR!
I'll have to look in more detail later, but my main concern is that this should probably be called `{slice,wide}Moves333` for now since it's hardcoded to 3x3x3.

**Comment by lgarron** [comment]
Alright, I've had a chance to try http://localhost:3333/experiments.cubing.net/cubing.js/play/?wideMoves333=true

I'm currently seeing:

- `L2` + `x2` → `L2 x2`
- `x` + `L` → `x L`
- `x` + `L` → `x L`
- `L` + `x` + `L` + `x` → `r r`

Would you be able to handle those cases in this PR as well?

**Comment by lgarron** [comment]
Also, similarly for http://localhost:3333/experiments.cubing.net/cubing.js/play/?sliceMoves333=true

- `L2'` + `R2` + `x2'` → `L2' R2 x2'`
- `x` + `L` + `R'` → `x L R'`
- `x` + `L` + `R'` → `x L R'`
- `R` + `x` + "L' → `l' L'` (`wideMoves` is not `true`)
- `L` + `R'` + `x` + `L` + `R'` + `x` → `M' M'`
- `L'` + `R` + `L'` + `R` + `x'` + `x'` → `L' R M x'` (This might happen if someone tries to input moves for `L2 R2 x2` if they press the keys for `L` and `R` simultaneously.)

**Comment by anicolao** [comment]
Thanks for the test cases. 

1. L2 + x2 → L2 x2 - I will fix this
2. x + L → x L - This is expected behaviour
3. x + L → x L - duplicate?
4. L + x + L + x → r r - this is expected behaviour as you didn't also pass `sameDirection=true`

For #2/#3 - there is no way to rotate the cube first and then execute the L move on the GAN cube. So I'm happy to handle the case but would want to add yet another flag to control when it applies, as my high-level goal here is to give the user the reconstruction of their solve as it was, not to simplify all cases. In some longer term fix, it makes sense to me to have `Alg.simplify` be able to do all of this, and have `experimentalAppendMove` be able to apply the subset that it wants to the tail of the alg, but I haven't thought through how to refactor it that way. Can you live with another flag for this out-of-order case, or come up with a way that a bluetooth cube might generate the sequence out of order?

For the next set:

1. L2' + R2 + x2' → L2' R2 x2' - will fix
2. x + L + R' → x L R' - same as #2/#3 above
3. R + x + L' → l' L' (wideMovesis nottrue`) - !!! impossible? I will fix
4. L + R' + x + L + R' + x → M' M' - needs `sameDirection=true`
5. L' + R + L' + R + x' + x' → L' R M x' - Looks correct to me

I will add all these test cases to the tests, so that we have a lasting record of what the expected behaviour is, but we probably need to agree on the expected behaviour to get to resolving these. (I can't help but note how easy it will be for me to put your test cases into my script to generate the tests, as opposed to someone else coming at this in a few months when I'm gone and having to edit the code laboriously to enter each of these cases :-)

**Comment by anicolao** [comment]
This boiled down to one bug: that if a slice or wide move was added, it would not also then get coalesced on to the previous alg. Fixed it by recursing. Added tests for all cases. 

However there are still some issues that we might need to discuss/agree on.

They are:

L2 + x2 → L2 x2 - I could not reproduce this. Test on line 140 passes.
x + L → x L - This is expected behaviour

and

L2' + R2 + x2' → L2' R2 x2' - could not reproduce. Test on line 263 passes.
R + x + L' → l' L' - could not reproduce. Test cluster on line 271 pass.

PTAL.

**Comment by anicolao** [comment]
Also when you get to looking at this again don't forget to respond to:

> I stopped short of pushing these options up to the element parameters so that we can check the shape of it before I change the public facing API. It isn't clear to me whether we'd prefer to keep supporting coalesce at the element level (it would mean sameDirection=oppositeDirection=true) and just add support for sameDirection and oppositeDirection or if we'd prefer to remove coalesce from the API and update all call sites.

**Comment by anicolao** [comment]
I rewrote history to take into account the new `identicalAlg` assert and added PR https://github.com/cubing/cubing.js/pull/213 so that can be committed while we finish off this PR.

**Comment by lgarron** [comment]
> L2 + x2 → L2 x2 - I could not reproduce this. Test on line 140 passes.

Hmm, I guess this is more like `L` + `L` + `x` + `x`, when I input at https://experiments.cubing.net/cubing.js/play/

I'm wondering if it might be better for us to figure out more specific edge cases, and for me to go to the drawing board about how to do this while hardcoding as little as possible.

Something like this:

```
interface PuzzleSpecificNotationInfo {
    axes: // ...
}

Alg.simplify({
  // ...
  puzzle: PuzzleSpecificNotationInfo | { notation: PuzzleSpecificNotationInfo }
})

interface PuzzleLoader {
  notation: PuzzleSpecificNotationInfo
}
```

For NxN, it would encode that axes include e.g. `R`, `2R`, ..., `2L'`, `L'`. But it could also work for Megaminx, or tetrahedral puzzles. This info could from from PuzzleGeometry, or be encoded by hand.

**Comment by anicolao** [comment]
You wrote: _I'm wondering if it might be better for us to figure out more specific edge cases, and for me to go to the drawing board about how to do this while hardcoding as little as possible._

I think the real problem is completely different: we don't agree on what "this" is. This pull request is about trying to make bluetooth cubes generate moves that correspond to what the user is doing on those cubes. The edge cases you are talking about are about how to make `Alg.simplify` do a good job of simplifying any alg into a concise notation / make use of slice or wide moves in the general case. Both things are desirable, but I have no interest in the `simplify` problem because it has no practical application for my use cases, while I am very interested in getting good solve reconstructions out of bluetooth cubes that can then be analyzed to enable the solver to improve.

In all universes, the bluetooth problem is different and more constrained than the general problem. If the general `Alg.simplify` did everything, possibly we could pass flags to constrain its behaviour. However it's far from it at the moment.

I've no problem withdrawing this pull request and building my own bluetooth implementation in my application instead, and that means we don't have to agree on how the bluetooth layer should work, and I can reuse the twisty player and alg code from the library while rewriting all the bluetooth code and how it interfaces with the rest. Does that sound better than getting this or a variation of it landed here?

**Comment by anicolao** [comment]
On the specific case of `L + L + x + x`, if executed on a physical cube, the potentially correct reconstructions in my view are one of `L2 x2` or `L r' x`.

**Comment by lgarron** [comment]
> I think the real problem is completely different: we don't agree on what "this" is. This pull request is about trying to make bluetooth cubes generate moves that correspond to what the user is doing on those cubes. The edge cases you are talking about are about how to make `Alg.simplify` do a good job of simplifying any alg into a concise notation / make use of slice or wide moves in the general case. Both things are desirable, but I have no interest in the `simplify` problem because it has no practical application for my use cases, while I am very interested in getting good solve reconstructions out of bluetooth cubes that can then be analyzed to enable the solver to improve.

Yeah, this is definitely something important. I really want to want to make sure that `cubing.js` has [least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), so it's really important to me that "append move" and `simplify` have compatible behaviour.

That said, one reason I've avoided tackling this problem is because we haven't had an active use case or test cases for it. But now that we have some test cases, it gives me an idea of what the implementation needs to support.

For your code, how tenable would it be to keep your own implementation of `experimentalAppendMove` for now?

https://github.com/cubing/cubing.js/blob/20cc84cbcd7e993e88a16c6f0d965970ed9bb394/src/cubing/twisty/model/TwistyPlayerModel.ts#L246-L267

The implementation can be refactored entirely using externally exposed functionality, and it would allow me to pursue a more general solution without blocking your work.

**Comment by anicolao** [comment]
The code I am working on that is currently in the library consists of this patchset to `experimentalAddMove` combined with a patch set that keeps track of the facing of the bluetooth cube by building quaternions for every orientation and finding the closest one, and applying a subset of likely `x`, `y` or `z` rotation in the GAN bluetooth code when it's clear the user has rotated the cube.

The quaternion code might be a good thing to land in the base library, but the bluetooth code in general needs refactoring and cleaning up, and so it seems simpler to me to just ignore it or fork it and write an entirely different bluetooth stack that includes the functionality currently in `experimentalAddMove` and that decouples our two problems completely. I've been trying from the get go to figure out if you will want these patches or not, because it really doesn't matter to me if you don't want them, but I am happy to contribute them if you do (see for examples the questions in https://github.com/cubing/cubing.js/issues/201). 

No doubt it'll be faster for me to just work on my own repository and only land here things that you want. So let me know if you want any of the orientation code or the bluetooth stack refactor I plan to do, and if so I can put up those PRs and see if they fit or not.

As for this PR we can just abandon it.

**Comment by lgarron** [comment]
Yeah, I'm grateful for the PR, but I've started working on a generalizable implementation: https://github.com/cubing/cubing.js/compare/puzzle-specific-alg-simplifications

Do I understand correctly that the following will be sufficiently useful for initial functionality?

- For any alg ending in a set of moves on the same axis:
  - If the net result of all those moves can be simplified into a single move (slice, wide, or rotation), replace them with that move.
  - Else, do nothing (for now).

This leaves room for improvement in the "else" case, but it will handle most algs with moves like `M'` or `r`.

**Comment by anicolao** [comment]
For my use case that rule is not correct ... it will modify moves that should not be modified when the goal is solve reconstruction. Here are examples that I think you'll want to handle differently than I will:

R L + L' = R L L' (_not_ R)
R L L' + L' = R L L2' (_not_ R L')
x L = x L (_not_ r)

In my mind the rule for `simplify` should be "reduce this alg to the shortest equivalent alg a solver can find". The reason I say so is that to me what `simplify` means is born out of the [language of computer algebra](https://en.wikipedia.org/wiki/Computer_algebra#Simplification), so by the principle of least surprise, simplify needs to mean that. If all simplify can do is cancel trivial sub-pieces of the string, like moves on a single axis, it probably needs a new name (like `coalesce`, which describes well the type of simplification you propose above).

This difference in our requirements is the core reason that it makes sense to have two implementations, because they are solving different problems. The problem of simplifying an alg is an interesting one, but it is never the right thing to do to a user's input if their goal is solve reconstruction. 

Both of these things have important use cases. Once the user has completed a solve, and has an accurate representation of the moves they made to do the solve, it's easy to imagine wanting to dice the solve into phases of a recognizable solution method (say roux or CFOP) and then for each phase apply `simplify` and suggest to the user shorter sequences that they might have used to achieve the sub-milestones of their solve (say first square in roux, or pair insertions in f2l in CFOP). A `simplify` routine that uses a solver to simplify would be fantastic for writing an automated solve critique for the user.

So my use cases are those two: reconstruct the user's actions so they have an accurate representation and playback of their solve; and simplify subsections of the user's solve in the computer algebra sense of the word in order to offer them better solutions. Possibly the simplification approach will need to be able to take inputs like "simplify, but within the constraint that the user will only be willing to do these moves { L, U, R, ... }". 

I do not have a use case for simplifying the algorithm in a single axis / simplifying moves that happen along a single axis. The closest I have is the `sameDirection` coalesce which simply compacts L + L = L2 and so on, but doesn't even combine L + L' (because if the user did that, they need to know they did it later). 

I should also note as an aside that _time_ is an important part of my solve reconstruction use case. I had not yet gotten as far with this patch to figure out when to _not_ simplify because of time, but it is clear that even if the user makes moves that logically do simplify like L + x = r, if L and x are not sufficiently proximate in the timeline, no simplification should be done.

**Comment by lgarron** [comment]
> R L + L' = R L L' (_not_ R)
> R L L' + L' = R L L2' (_not_ R L')

Ah, right. But in this case, this would be solved by requiring the same direction for all moves in an axis grouping that have the same quantum move (e.g. `L` and L2` can be in the same group, but not `L` and `L2'`?
(e.g. using a `sameDirection` param)

> x L = x L (_not_ r)

I'm having trouble following here. Is the idea that `x` is assumed to be a rotation before a given solution step, and `L` the first move of a solution step?

As far as I'm concerned, `x L` vs. `L x` is a transcription race condition. Both are equally plausible results of keyboard input, and I don't want to handle Bluetooth input or reconstructions differently.

> Possibly the simplification approach will need to be able to take inputs like "simplify, but within the constraint that the user will only be willing to do these moves { L, U, R, ... }".

I think we'll definitely need this at some point, particularly for "rewrite this alg without moving any centers".

> I had not yet gotten as far with this patch to figure out when to _not_ simplify because of time, but it is clear that even if the user makes moves that logically do simplify like L + x = r, if L and x are not sufficiently proximate in the timeline, no simplification should be done.

One option here might be to insert `Pause` leaves/comments/newlines whenever you want to prevent simplification across certain ranges of an alg.

**Comment by anicolao** [comment]
Yes, I have use for a `sameDirection` version so if that option is there then your approach is a big step closer to being usable for me.

Regarding x L vs L x, and whether or not keyboard vs bluetooth input vs reconstructions are different, we still have a disconnect. My goal is to make the alg match what the user did on their bluetooth cube. If the cube produces x L as its output, the user must have rotated first, and then performed a move. There is no race condition in my code that produces this output and no ambiguity about what the user did.

It might be helpful to be a bit pedantic about what each thing means:

1. Keyboard input: the user is manipulating the on screen cube via keyboard shortcuts
2. Bluetooth input: the user is manipulating the on screen cube using the physical cube as a game controller
3. Reconstructions: the user is manipulating and looking at the physical bluetooth cube, and expects to later be able to replay what their physical actions were

In my mind, the two input cases should not contain any race conditions. The user is looking at the screen and can see what they're doing, and they expect the screen to match what they've done. 

If they're using a keyboard, no rewriting should ever be done in my opinion. The input is unambiguous.

If they're using a cube, and they do an M slice move and the on screen display does a three move animation that is simply wrong. If they do a cube rotation and then a face move, and the screen shows a wide move, that is simply wrong. Since the user can choose to do x L, the system has to be able to display it.

If they're watching a reconstruction, ideally it would exactly match their physical actions, but it's easier to accept some inaccuracies.

You say you don't want to treat the cases differently but I am not really clear if that's true. If you're willing to have a `sameDirection` flag, that will let the would-be application author get more of what they want out of your solution. From my PoV the problem with the approach you are exploring is that you've chosen to view two operations that commute as equivalent and consider it a 'race condition' as to which order they occur in, while I view it as the user provided an input and the only source of ambiguity is whether or not the sensors on a bluetooth cube pick up one face move before the other.

On the other hand, if it is really true that you don't want to treat the cases differently, then I think you're in the classic trap of a library designer or bottom-up designer, where you want to apply an elegant solution to a problem that doesn't exist. I just can't imagine anyone who wants their keyboard input rewritten at all; nor anyone who wants the bluetooth cube input to be animated incorrectly onscreen; nor anyone who wants to be shown a simplified algorithm being satisfied with one that can only handle moves in one axis. Since there are three distinct use cases, with different requirements, a single solution is not going to cut it: minimally you need some mess of flags that lets the caller decide what codepath you will execute.

**Comment by anicolao** [comment]
I'd like to close off / clean up any cruft I have left behind, so I am going to close this as a dead end.

**Comment by lgarron** [comment]
Yeah, I'm sorry this didn't quite get there. I still have the `puzzle-specific-alg-simplifications` branch, which I want to keep working on how now that I've cleaned up some of the project logistics.

**Comment by anicolao** [comment]
No problem! If your branch gets merged, I'll take a look and see if it covers my use case or not.

**Comment by lgarron** [comment]
> If your branch gets merged, I'll take a look and see if it covers my use case or not.

Alright, the new implementation is fairly stable on `main`. Could you check if it works for you?

```ts

import { Alg, experimentalAppendMove, Move } from "cubing/alg";
import { cube3x3x3 } from "cubing/puzzles";

const alg = new Alg("R U R2 R' M'");
alg.experimentalSimplify({cancel: true}).log(); // R U R M'
alg.experimentalSimplify({cancel: true, puzzleLoader: cube3x3x3}).log(); // R U r
experimentalAppendMove(alg, new Move("L'"), {cancel: true, puzzleLoader: cube3x3x3}).log() // R U x
experimentalAppendMove(alg, new Move("M"), {cancel: true, puzzleLoader: cube3x3x3}).log() // R U R
experimentalAppendMove(alg, new Move("M"), {cancel: {directional: "same-direction"}, puzzleLoader: cube3x3x3}).log() // R U R2 R' M' M
experimentalAppendMove(alg, new Move("M3'"), {cancel: {directional: "same-direction"}, puzzleLoader: cube3x3x3}).log() // R U R2 R'
experimentalAppendMove(alg, new Move("M3'"), {cancel: {directional: "same-direction", puzzleSpecificModWrap: "none"}, puzzleLoader: cube3x3x3}).log() // R U R2 R' M4'
```

**Comment by anicolao** [comment]
> Alright, the new implementation is fairly stable on `main`. Could you check if it works for you?

I didn't actually try it out because your examples are sufficient for me to know I can't use it.

> ```ts
> const alg = new Alg("R U R2 R' M'");
> 
> alg.experimentalSimplify({cancel: true}).log();
> // R U R M'
```

When doing solve reconstruction, I do not want to cancel opposite moves.

```
> new Alg("R U R2 R' M' L'").experimentalSimplify({cancel: true, puzzleLoader: cube3x3x3}).log();
> // R U x
```

Same problem, but worse because this also combined moves that were made separately.

```
> experimentalAppendMove(alg, new Move("L'"), {cancel: true, puzzleLoader: cube3x3x3}).log()
> // R U x
```

I'll never want a cube rotation unless the user physically rotated their bluetooth cube. OTOH, I will _always_ want a cube rotation if the user physically rotated their cube, even if they return to the original orientation with something like `x x'`.


```
> experimentalAppendMove(alg, new Move("M"), {cancel: {directional: "same-direction"}, puzzleLoader: cube3x3x3}).log()
> // R U R2 R' M' M
```
This is good.
```
> experimentalAppendMove(alg, new Move("M3'"), {cancel: {directional: "same-direction"}, puzzleLoader: cube3x3x3}).log()
> // R U R2 R'
```
This is bad because it eliminated moves the user actually made.
```
> experimentalAppendMove(alg, new Move("M3'"), {cancel: {directional: "same-direction", puzzleSpecificModWrap: "none"}, puzzleLoader: cube3x3x3}).log()
> // R U R2 R' M4'
> 
```
This is good.

Long story short: it looks like all the simplifications that are possible I don't want, and there are no examples here of simplifications that I do want. Does the patch introduce wide and slice moves when the corresponding single face moves plus cube rotations happen in a short time period (like < 100ms)?

**Comment by lgarron** [comment]
Thanks, that's helpful to know! Could you provide real-world-ish examples for when `{directional: "same-direction", puzzleSpecificModWrap: "none"}, puzzleLoader: cube3x3x3}` provides the wrong results, and what you'd like to see instead? So far I can discern:

- Avoid introducing full-puzzle rotations if they weren't already there. So `r L'` → `r L'` rather than `x`.

I'd be happy to add a "avoidIntroducingRotations` option for that.

> Does the patch introduce wide and slice moves when the corresponding single face moves plus cube rotations happen in a short time period (like < 100ms)?

I want to be able to handle this in a streaming protocol eventually, but the player is not a place I can introduce this yet. For now, I would add a `Newline`, `LineComment`, or `Pause` between any two moves that you want to prevent from cancelling, and then take them out when "pretty-printing".

**Comment by anicolao** [comment]
For solve reconstruction, my goal is to provide the user with a visualization and analysis of their actual solve. So same-direction+no mod will be close a lot of the time, as it'll do nothing if it doesn't do wide, slice, or rotation moves; all that is left is to merge adjacent moves of the same face. That'll mostly be OK (though ideally I could provide a time interval and it would avoid merging R R -> R2 if the timestamps are too far apart, by which I mean the timestamp of the _end_ of the first R versus the _start_ of the next R). 

The real bottom line though is that this just isn't the behaviour I want out of `simplify` at all. What `simplify` should really be used for is solve _analysis_, not solve _reconstruction_. So I want an append move that does what I said above, and a simplify that [does things like this:](https://alpha.twizzle.net/edit/?alg=L+D%27+R2+B+R%27+F%27+L+R%27+D+%2F%2F+First+block%289%29+vs+%2810%29+y+x2++x%27++U2+R%27+B+D+R+F+B&setup-alg=R+U%27+D%27+R2+F+U%27+B+R+L+F2+R2+F%27+L2%27+U2+F%27+L2%27+U2+B2%27+D2%27+F%27+D2%27)

```
R U' D' R2 F U' B R L F2 R2 F' L2' U2 F' L2' U2 B2' D2' F' D2' // Scramble
L D' R2 B R' F' L R' D  // First block(9) vs (10) y x2  x'  U2 R' B D R F B
```

So you see in this case my simplifier is doing a lot of things your simplifier does not seem capable of: ignoring cube rotations that are part of inspection, finding a 7 move solution instead of a 9 move solution for the green FB with white on the bottom. At the same time it's not doing whatever it wants - it is simplifying the user's chosen first block, not arbitrarily picking the best FB (which is orange with yellow bottom: `x' U2 r B' r2 D F D2`). This is what it means to `simplify`.

So basically the code here does either too much or too little in all the dimensions I care about: for solve reconstruction, it's doing a bunch of simplifications I don't want and basically nothing that I do want; and for solve analysis it's not even close.

As I have both implemented already outside the library perhaps my use cases just don't matter here. I've already tried to express that this approach of simplifying same-axis moves might be very nice and general across different twisty puzzles, but it solves a problem I don't have, namely the simplifying that results is way too weak, and the move coalescing that results is way too strong.

**Comment by anicolao** [comment]
On the subject of inserting `NewLine` or `LineComment` to prevent coalescing, it's approximately as much work to do that as to not use this functionality at all and just do the coalescing myself. In that way I can easily deal with timestamps and other features I need that are not here yet, and that don't seem likely to get added to your library based on my track record so far, and are easy enough for me to implement outside. What I need in the fullness of time for solve reconstruction is to have start and end timestamps for every move, coalescing that is timestamp aware, and the ability to score things like corner cutting or aggregate idle time to show the user their most egregious recognition pauses / train them on recognition situations where they are weak.

So going to the trouble of adding special code to defeat your code to get a tiny bit of essentially trivial move coalescing doesn't seem like a good work/reward ratio, and even more so if the code that's in this patch is in `simplify` which seems very much the wrong interface for it to me and thus unlikely to remain working over time.

**Comment by lgarron** [comment]
Thanks, those thoughts are still useful to hear!

I think we're going to continue disagree on what it means to simplify an alg. What you describe sounds more like the kind of thing we want to tackle with `cubing/search`. But it's certainly something I'll keep in mind before I move the functionality out of "experimental".

As for timestamp-based animations & cancellations, as well as alg scoring, this is *definitely* something I want to have solid support for — you can find traces of setup work in the project history. Unfortunately the project is still too young to address most of those needs, but you may be interested to play around with https://github.com/cubing/cubing.js/blob/0ec26d5324afd93b26181a94ac3511cff5da24cf/src/cubing/twisty/controllers/indexer/simultaneous-moves/SimultaneousMoveIndexer.ts#L18-L41 for per-move animation timestamps.
